File: backend/app/schemas/cotacao.py
from datetime import datetime

from pydantic import BaseModel, Field


class Endereco(BaseModel):
    cep: str
    cidade: str
    uf: str = Field(min_length=2, max_length=2)


class VolumeIn(BaseModel):
    quantidade: int = Field(gt=0)
    comprimento_cm: float = Field(gt=0)
    largura_cm: float = Field(gt=0)
    altura_cm: float = Field(gt=0)
    peso_kg: float = Field(gt=0)


class CotacaoCreate(BaseModel):
    origem: Endereco
    destino: Endereco
    valor_nf: float = Field(gt=0)
    peso: float = Field(gt=0)
    volumes: list[VolumeIn]
    transportadoras_ids: list[str] | None = None  # None = todas as ativas


class ErroResultado(BaseModel):
    codigo: str
    mensagem: str


class ResultadoTransportadora(BaseModel):
    transportadora_id: str
    transportadora: str
    status: str  # success | error | timeout | processing
    valor_frete: float | None = None
    prazo_dias: int | None = None
    moeda: str = "BRL"
    erro: ErroResultado | None = None
    request_id: str


class CotacaoOut(BaseModel):
    id: str
    status: str  # processing | completed | completed_with_errors | failed
    cubagem_m3: float
    melhor_opcao_id: str | None = None
    resultados: list[ResultadoTransportadora] = []


class CotacaoListItem(BaseModel):
    id: str
    status: str
    origem_cep: str
    origem_cidade: str
    origem_uf: str
    destino_cep: str
    destino_cidade: str
    destino_uf: str
    valor_nf: float
    peso: float
    cubagem_m3: float
    melhor_frete: float | None = None
    transportadora_id: str | None = None
    transportadora: str | None = None
    prazo_dias: int | None = None
    total_resultados: int = 0
    resultados_sucesso: int = 0
    created_at: datetime


class CotacaoListaResponse(BaseModel):
    items: list[CotacaoListItem]
    total: int
    page: int
    page_size: int
    total_pages: int


class SelecionarTransportadora(BaseModel):
    transportadora_id: str

File: frontend/src/types/cotacao.ts
export interface Endereco {
  cep: string;
  cidade: string;
  uf: string;
}

export interface VolumeIn {
  quantidade: number;
  comprimento_cm: number;
  largura_cm: number;
  altura_cm: number;
  peso_kg: number;
}

export interface CotacaoCreate {
  origem: Endereco;
  destino: Endereco;
  valor_nf: number;
  peso: number;
  volumes: VolumeIn[];
  documento_destinatario?: string | null;
  transportadoras_ids?: string[] | null;
}

export interface ErroResultado {
  codigo: string;
  mensagem: string;
}

export type StatusResultado = "processing" | "success" | "error" | "timeout";
export type StatusCotacao = "processing" | "completed" | "completed_with_errors" | "failed";

export interface ResultadoTransportadora {
  transportadora_id: string;
  transportadora: string;
  status: StatusResultado;
  valor_frete: number | null;
  prazo_dias: number | null;
  moeda: string;
  erro: ErroResultado | null;
  request_id: string;
  detalhamento?: {
    regiao_tarifaria?: string;
    peso_real_kg?: number;
    peso_cubado_kg?: number;
    peso_considerado_kg?: number;
    prazo_dias?: number;
    faixa?: { from_kg: number; to_kg: number };
    taxas_detalhadas?: Array<{ tipo: string; valor: number | null }>;
  } | null;
}

export interface CotacaoOut {
  id: string;
  status: StatusCotacao;
  cubagem_m3: number;
  melhor_opcao_id: string | null;
  recomendacao_motivo?: string | null;
  resultados: ResultadoTransportadora[];
}

export interface CotacaoListItem {
  id: string;
  status: StatusCotacao;
  origem_cep: string;
  origem_cidade: string;
  origem_uf: string;
  destino_cep: string;
  destino_cidade: string;
  destino_uf: string;
  valor_nf: number;
  peso: number;
  cubagem_m3: number;
  melhor_frete: number | null;
  transportadora_id: string | null;
  transportadora: string | null;
  prazo_dias: number | null;
  total_resultados: number;
  resultados_sucesso: number;
  created_at: string;
}

export interface CotacaoFiltros {
  page: number;
  page_size: number;
  status?: string;
  origem_uf?: string;
  destino_uf?: string;
  transportadora_id?: string;
  data_inicio?: string;
  data_fim?: string;
  busca?: string;
}

export interface CotacaoListaResponse {
  items: CotacaoListItem[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}
